typings/commander-tests.ts

```diff
@@ -37,6 +37,12 @@ function collect(val: string, memo: string[]) {
     memo.push(val);
     return memo;
 }
+/*
```

Comment:
I was wondering about the void and had to look at the code to realise why it is there. The coercion routine can return void to indicate the coercion failed and the default value should be used. e.g.

```
function shortString(value: string) {
   if (value.length < 4)
      return value;
}

program
   .option('-s <str>', 'string less than four characters long', shortString, '...');
```

Reply:
allowing void for me allow using shorter code:
```javascript
function collect2(val: string, memo: string[]) { memo.push(val);}
// instead of
function collect2(val: string, memo: string[]) { memo.push(val); return memo;}
// of cours this last version should be sorted, but not working, push returning a length
const collect2 = (val: string, memo: string[]) => memo.push(val);
```
so prefer keep it as it, since, no alternative have real advantage.
the current version can store collection in readonly collection.
or also make the following collector fonctionnal:
```javascript
const collect2 = (val: string, memo: string[]) => [...memo, val ];
```

Reply:
I read the code and not sure of existing behaviour, so I tried it.

I do not think this example is working as you might expect. And to be fair, my own example did not work as I assumed either! The Commander code tests the function return value against null (for some reason) but a void function returns undefined. 

After that, I am not sure under what circumstances a void function makes sense for the coercion routine, if any.

```
const program = require('commander');

function collect2(val, memo) {
  memo.push(val);
}

program
  .option('-c,collect <value>', 'A repeatable value void return', collect2, []);

program.parse(process.argv);

console.log(program.collect);
```

```
$ node void.js                    
[]
$ node void.js -c a          
undefined
$ node void.js -c a -c b
[ 'a', 'b' ]
$ node void.js -c a -c b -c c
undefined
$ node void.js -c a -c b -c c -c d
[ 'a', 'b', 'c', 'd' ]
```

Reply:
So, I should had I read the implementation.

Is the current implementation is logic ?
I can change It, to make it match my last typing.

I'm not sure I understood the way you expect it to works.

Reply:
I am interested in improving the TypeScript to better match the implementation and the examples in the README.

I do not want to change the implementation. In particular, I don't want to make the example from your original post work (sorry):
```
    .option('--hostname <hostname>', 'hostname used for binding http server', (value: string, previous: string[]) => { previous.push(value); }, [] as string[])
```

Are you still interested in improving the types without changing the code? No problem if are not interested in just the types, I will open an issue or PR with some of what we learnt and your suggestions.

I have learnt a lot looking into this PR!